src/bootstrap/doc.rs

```diff
@@ -650,21 +651,22 @@ macro_rules! tool_doc {
                 let target = self.target;
                 builder.info(&format!("Documenting stage{} {} ({})", stage, stringify!($tool).to_lowercase(), target));
 
-                // This is the intended out directory for compiler documentation.
-                let out = builder.compiler_doc_out(target);
-                t!(fs::create_dir_all(&out));
-
-                let compiler = builder.compiler(stage, builder.config.build);
-
                 if !builder.config.compiler_docs {
                     builder.info("\tskipping - compiler/tool docs disabled");
                     return;
                 }
 
+                // This is the intended out directory for compiler documentation.
+                let out = builder.compiler_doc_out(target);
+                t!(fs::create_dir_all(&out));
+
                 // Build rustc docs so that we generate relative links.
                 builder.ensure(Rustc { stage, target });
+                // Rustdoc needs the rustc sysroot available to build.
+                builder.ensure(check::Rustc { target: builder.config.build });
 
                 // Symlink compiler docs to the output directory of rustdoc documentation.
+                let compiler = builder.compiler(stage, builder.config.build);
```

Comment:
Stage handling is wrong here, right? If we hit this code from some entry path with stage != builder.top_stage, then we'll not have the right sysroots in the right places after the check::Rustc above.

We either need to only support documenting in the top stage (potentially reasonable, not sure there's a reason to do so in other stages) or add support for arbitrary stages to the check code (seems like a big hassle, likely to prove quite difficult).

Reply:
> We either need to only support documenting in the top stage (potentially reasonable, not sure there's a reason to do so in other stages) 

:+1: makes sense, I'll add an assert.

Reply:
An assert seems wrong. We can statically ensure this by dropping the stage field from these structs and using the top stage within.

Reply:
Oh, good point. Done.